### PR TITLE
NIAD-1720: MedicationRequest: GPC Spec change (unpublished) - requester override recorder as attribution

### PR DESCRIPTION
**Pull request by bbell-kainos**

### CODE REVIEW

File: service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementMapperTest.java [hunk 1]
```diff
@@ -199,13 +175,6 @@ public void When_MappingObservationJson_Expect_NarrativeStatementXmlOutput(Strin
             Arguments.of(INPUT_JSON_WITH_PLAN_NO_OPTIONAL_FIELDS, OUTPUT_XML_WITH_AUTHORISE_NO_OPTIONAL_FIELDS),
             Arguments.of(INPUT_JSON_WITH_PLAN_NO_STATUS_REASON_CODE, OUTPUT_XML_WITH_AUTHORISE_DEFAULT_STATUS_REASON_CODE),
             Arguments.of(INPUT_JSON_WITH_PLAN_NO_INFO_PRESCRIPTION_TEXT, OUTPUT_XML_WITH_AUTHORISE_REPEAT_PRESCRIPTION),
-            Arguments.of(INPUT_JSON_WITH_ORDER_NO_REQUESTER, OUTPUT_XML_WITH_AUTHORISE_NO_ORG_INFO),
-            Arguments.of(INPUT_JSON_WITH_ORDER_ON_BEHALF_OF_NOTHING, OUTPUT_XML_WITH_AUTHORISE_NO_ORG_INFO),
-            Arguments.of(INPUT_JSON_WITH_ORDER_ON_BEHALF_OF_ORG, OUTPUT_XML_WITH_AUTHORISE_ORG_INFO),
-            Arguments.of(INPUT_JSON_WITH_PLAN_NO_REQUESTER, OUTPUT_XML_WITH_PRESCRIBE_NO_ORG_INFO),
-            Arguments.of(INPUT_JSON_WITH_PLAN_ON_BEHALF_OF_NOTHING, OUTPUT_XML_WITH_PRESCRIBE_NO_ORG_INFO),
-            Arguments.of(INPUT_JSON_WITH_PLAN_ON_BEHALF_OF_ORG, OUTPUT_XML_WITH_PRESCRIBE_ORG_INFO),
-            Arguments.of(INPUT_JSON_ON_BEHALF_OF_ORG_WITH_MULTIPLE_TYPES, OUTPUT_XML_WITH_AUTHORISE_ORG_INFO_MULTIPLE_TYPES),
```
Comment (ElliotB-96) [inline service/src/test/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementMapperTest.java:208]: Are there no additional tests to be added for this change?

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementMapper.java [hunk 2]
```diff
@@ -296,7 +269,20 @@ private String buildParticipant(MedicationRequest medicationRequest) {
         var isOrganization = buildPredicateReferenceIsA(ResourceType.Organization);
         Predicate<Reference> isRelevant = isPractitioner.or(isPractitionerRole).or(isOrganization);
 
-        if (medicationRequest.hasRecorder() && medicationRequest.getRecorder().hasReference()) {
+        if (medicationRequest.hasRequester()) {
+            final var requester = medicationRequest.getRequester();
+            final var agent = requester.getAgent();
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementMapper.java:274]: You check for `if (medicationRequest.hasRequester()) {` but not for `requester.hasAgent()` ?

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementMapper.java [hunk 3]
```diff
@@ -296,7 +269,20 @@ private String buildParticipant(MedicationRequest medicationRequest) {
...
+
+            if (ResourceType.Practitioner.name().equals(agent.getResource().getIdElement().getResourceType())
+                && requester.hasOnBehalfOf()) {
```
Comment (bartek-sarul) [inline service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementMapper.java:277]: Might be more readable if you would add more indentation on this line

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementMapper.java (hunk 3 above)
Comment (bartek-sarul) [reply service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/MedicationStatementMapper.java:278]: What I meant here is:
```
if (ResourceType.Practitioner.name().equals(agent.getReferenceElement().getResourceType())
        && requester.hasOnBehalfOf()) {
    var onBehalfOf = requester.getOnBehalfOf();
    ...
```

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

**Pull request merged**